src/main/java/com/eerussianguy/blazemap/engine/cache/RegionMDCache.java

```diff
@@ -1,67 +1,76 @@
 package com.eerussianguy.blazemap.engine.cache;
 
 import java.io.IOException;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.Iterator;
+
 
 import net.minecraft.world.level.ChunkPos;
 
 import com.eerussianguy.blazemap.BlazeMap;
 import com.eerussianguy.blazemap.api.util.MinecraftStreams;
 import com.eerussianguy.blazemap.api.util.RegionPos;
 import com.eerussianguy.blazemap.engine.async.DebouncingDomain;
-import com.eerussianguy.blazemap.engine.async.PriorityLock;
 import com.eerussianguy.blazemap.profiling.Profilers;
 
 public class RegionMDCache {
     private static final int CHUNKS = 0x20;
     private static final byte VOID = 0x00, DATA = 0x01;
 
-    private final ChunkMDCache.Persisted[] chunks = new ChunkMDCache.Persisted[CHUNKS * CHUNKS];
+    private final ChunkCacheStore chunks = new ChunkCacheStore(CHUNKS * CHUNKS);
     private final DebouncingDomain<RegionMDCache> debouncer;
-    private final PriorityLock lock = new PriorityLock();
     private final RegionPos pos;
-    private boolean dirty;
+
+    // These are both invoked at the LevelMDCache level, due to the movement between buffer and file
+    public final ReentrantReadWriteLock fileLock = new ReentrantReadWriteLock();
+    public final ReentrantLock bufferLock = new ReentrantLock();
 
     public RegionMDCache(RegionPos pos, DebouncingDomain<RegionMDCache> debouncer) {
         this.pos = pos;
         this.debouncer = debouncer;
     }
 
-    PriorityLock lock() {
-        return lock;
-    }
-
     public boolean isDirty() {
-        return dirty;
+        for (ChunkMDCache.Persisted chunk : chunks) {
+            if (chunk != null && chunk.isDirty()) {
+                return true;
+            }
+        }
```

Comment:
I have issues.
On one hand, this is correct.
On the other hand, Chunk caches were passing their dirty state to their parent Region cache which was being kept locally so that we don't need to check 1024 chunk caches every time... that's why it was there in the first place.

Now, I could be wrong about the impact of these approaches, but my educated guess was that it'd be more performant to just keep a local value at the region level.

Reply:
This was a trade off I did spend time looking into, because I really _didn't_ want to have to read 1024 chunks. But the trade-off I found was worth it.

Readin back on my notes from the time (https://discord.com/channels/709852148170620948/1267059848705605673/1267530673053241384), TL;DR:
>  I think minimising frequent thread synchronisation is more important than the occasional check for cleanliness before writing to a file max every 30s

The longer version:
The problem here is that we're dealing with three different optimisation goals here, and it's only possible to satisfy 2 of the 3:
- Best chunk access performance
- Best dirty check performance
- Thread safety

Previously, the code was optimising for "dirty check performance" and "thread safety". However, this was at the cost of at the cost of "chunk access performance". This is because, if _any_ chunk was being accessed, **every other chunk had to wait for it to be done**. Normally this is fine as the operation is generally fast, but it was _the_ major cause of the hitching if one thread tried to load a chunk from a region that needed to be read from disk. But the dirty check was fast, and it was always accurate as no chunks could be changed while writing them to disk.

In the new version of the code, I changed the locks so each chunk was locked separately. This meant that the game only hitched if _every_ thread was waiting for a chunk that was waiting on disk IO, as otherwise the threads that were looking up chunks in already loaded regions were not impacted at all by the other threads waiting on IO.

However, when it came to the `dirty` check, I then had to optimise between "thread safety" or "dirty check performance". _If_ I were to similarly use a variable on the region to optimise for "dirty check performance", this _could not_ be thread safe:
- When writing to disk, each chunk is only locked while it personally is being written. That means the chunk can be changed _after_ it's written to disk. If the `dirty` check is held on the region and applied after the file is written to disk, the updates to that chunk will _not_ be recognised. 
- Alternatively, if all chunks are locked while writing to disk to make sure that `dirty` check is accurate at the end of writing, that again causes unnecesary synchronisation on both reads and writes.
- If instead the variable is updated _before_ writing, to allow it to be updated by any chunks changed after they've been written out, this still won't be accurate if a chunk is changed before the write process gets to it but is then still output in its final state to disk. 
- Additionally, if the variable is updated before writing, it will be left as `clean` if an exception causes the write to fail, again not reflecting the true status of the region.
- I did even think of using the bits in a `long` to represent each chunk, with `1` meaning dirty and `0` meaning clean. That way, `dirty >  0` means the chunk is dirty. But it would have needed multiple `long`s to work, and those longs would have had to be synchronised, and it wasn't a clean solution.

Alternatively, I looked at where that dirty check was taking place. The dirty check is _only_ called before writing the region MD to file. This should happen at most every 30s, and should be only on the data threads. Given the rarity of this check and the fact it shouldn't impact the render thread, I considered it the least important of the three optimisation variables in the trade off. 

But also, that's why the complication around the locking state of the `dirty` variable on each chunk. Yes the region has to read 1024 variables, but it _never_ has to wait for a lock. The variable is `volatile`, so its value should be up to date in the CPU cache before reading. This is as fast as I can make accessing the `dirty` vairable in all chunks from a region while still protecting its access behind a function call (so it can't be written to without a lock).

In other words, I had to make a trade off, and I chose the set of trade offs that maximised average performance and correctness. And that's why it takes longer to check if a region is dirty in this new code.